### ISSUE TITLE
[java] CloseResource: false positive with reassignment detection

### ISSUE THREAD

**Issue by djh82** [pmd/pmd#2977]
Rule: [CloseResource](https://pmd.github.io/latest/pmd_rules_java_errorprone.html#closeresource)

Upgraded to 6.30.0, having previously been using 6.28.0.  Now seeing false positives with the CloseResource rule, with this contrived example:

```java
import java.io.RandomAccessFile;

public class FalsePositive {
    private boolean isHundredBytes = false;

    public void bar() throws Exception {
        RandomAccessFile file = null;
        try {
            file = new RandomAccessFile("name", "r");
            this.isHundredBytes = file.length() == 100;  /* violation here... */
        } finally {
            file.close();
        }
    }
}
```

**Comment by davidhart82** [comment]
Assume it was introduced by #2811

**Comment by dhalperi** [comment]
More false positives, with streams: https://github.com/batfish/batfish/blob/614c3bf2178044263b26a94d18e72c1cf221dba1/projects/batfish/src/main/java/org/batfish/bddreachability/BDDReachabilityAnalysisFactory.java#L1496
https://github.com/batfish/batfish/blob/master/projects/coordinator/src/main/java/org/batfish/coordinator/WorkMgr.java#L2156-L2159


We follow this pattern fairly commonly:
```java
Stream<T> stream = ...;
if (condition) {
    stream = stream.filter(...);
}
```

all these are giving false positives in the if.

**Comment by oowekyala** [comment]
The original comment appears to be fixed by #3061, but not the stream ones... I'll look into this

**Comment by djh82** [comment]
I'm not sure this has been fixed; still seeing issues with 6.31.0

**Comment by Kittera** [comment]
> I'm not sure this has been fixed; still seeing issues with 6.31.0

I'm seeing this myself on 6.31.0, just updated in order to test. Still flagging it even if I immediately close the Scanner.
![image](https://user-images.githubusercontent.com/5945284/108310477-4b5f3400-7168-11eb-853a-1fd13efc8d1c.png)
![image](https://user-images.githubusercontent.com/5945284/108310708-b01a8e80-7168-11eb-9e01-a41c884ec95d.png)

**Comment by oowekyala** [comment]
Thanks for your feedback I'll try and see what went wrong

**Comment by oowekyala** [comment]
@djh82 Can you provide a code sample which includes the context around those statements (method & class declaration), and does not use custom types like your File? Even going back to pmd 6.30, I cannot reproduce the issue

@Kittera This looks like something else. Does the violation also occur if you tick the first checkbox in the configuration panel?

**Comment by Kittera** [comment]
> @Kittera This looks like something else. Does the violation also occur if you tick the first checkbox in the configuration panel?

@oowekyala Yes, as well as producing additional warnings about any `close()` statements not within a finally-block. Except, conspicuously, for the ones pictured below.

In my case I don't necessarily want to close it within a finally-block for this method anyway, as the associated `try` is inside a `do-while` prompting loop and I don't want to break the connection to the keyboard input. I also use the Scanner not only to capture a filename but, in the event of a `FileNotFoundException`, to then ask the user whether to quit or try again, which would be complicated by trying to use a try-with-resources instead due to scope mechanics.

Even if I put a `console.close()` at the end of the method, where I'd definitely be done with it in all scenarios, the warning is still produced. I have several `close()` statements at various points in this method and none of them seem to satisfy.

![image](https://user-images.githubusercontent.com/5945284/108588767-beef7580-730f-11eb-88f5-7c6e19898d84.png)

The other two items are results of unfinished development in other areas of this class. They are not from PMD.

**Comment by adangel** [comment]
@djh82 Your code sample does not compile. If I use RandomAccessFile (like you used initially), PMD works fine and no CloseResource violation is reported.
Without a complete reproduction case, we can't fix anything.... Please provide more context.

@Kittera I don't think, this is a false positive and the rule is correct. There might be cases, where `console` is not closed: E.g. there can exceptions/errors happen at line 87, 90, 92 - which means console is not closed as you wanted.
Therefore the only way to be sure, to close console, is, to use try (either with finally and manual closing or with resources).

Here's your original method and 2 ways to fix it:

```java
package mypackage;

import java.io.*;
import java.util.*;

public class ScannerFalsePositive {

    public static Scanner promptOriginal() {
        boolean valid = false;

        final Scanner console = new Scanner(System.in);
        Scanner result = null;
        do {
            System.out.println("Please enter filename: ");
            try {
                result = new Scanner(new File(console.next()));
                valid = true;
            } catch (final FileNotFoundException oops) {
                System.out.println("File not found, try again?");
                
                if (console.next().equalsIgnoreCase("Y")) {
                    console.close();
                    System.exit(0);
                }
            }
        } while (!valid);
        
        console.close();

        return result;
    }

    public static Scanner promptTryFinally() {
        boolean valid = false;

        Scanner result = null;
        Scanner console = null;
        try {
            console = new Scanner(System.in);
            do {
                System.out.println("Please enter filename: ");
                try {
                    result = new Scanner(new File(console.next()));
                    valid = true;
                } catch (final FileNotFoundException oops) {
                    System.out.println("File not found, try again?");
                    
                    if (console.next().equalsIgnoreCase("Y")) {
                        console.close();
                        System.exit(0);
                    }
                }
            } while (!valid);
        } finally {
            if (console != null) {
                console.close();
            }
        }

        return result;
    }

    public static Scanner promptTryResources() {
        boolean valid = false;
        Scanner result = null;

        try (Scanner console = new Scanner(System.in)) {
            do {
                System.out.println("Please enter filename: ");
                try {
                    result = new Scanner(new File(console.next()));
                    valid = true;
                } catch (final FileNotFoundException oops) {
                    System.out.println("File not found, try again?");
                    
                    if (console.next().equalsIgnoreCase("Y")) {
                        console.close();
                        System.exit(0);
                    }
                }
            } while (!valid);
        }

        return result;
    }
}
```

Oh and should be the check for "Y"/"N" inverted? It looks right now, that you exit, when you answer the question "try again" with Y...

**Comment by djh82** [comment]
@adangel weird, don't know why it changed from RandomAccessFile, since that is what we're using. Will try and get a small and succinct reproducer for you tomorrow.

**Comment by djh82** [comment]
@adangel double checked our code, and there were some slight differences.  Have updated the example and tested this morning.  Ideally we'd be using try-with-resources, but that's currently not available to us!

**Comment by adangel** [comment]
@djh82 Thanks, I can reproduce it now. It has something to do with declaring the variable and re-assigning it.

One possible workaround could be (depends on your actual code) to use two try-blocks:

```java
    public void bar3() throws Exception {
        try {
            RandomAccessFile file = new RandomAccessFile("name", "r");
            try {
                boolean isHundredBytes = file.length() == 100;
            } catch (IOException e) {
                // no-op
            } finally {
                file.close();
            }
        } catch (IOException e) {
            // ignored
        }
    }
```

**Comment by djh82** [comment]
@adangel excellent.  Yes, while digging out the code and writing that reproducer, I did _fix_ the error by doing exactly that.  Though I think the pattern is the usual paradigm when try-wth-resources is not available (usually with a null-guard and try/catch in the finally too).

**Comment by adangel** [comment]
@djh82 Unfortunately I looked at the wrong file, I still had the ScannerFalsePositive sample in my test project which triggered a violation, but your file runs fine.
So I still can't reproduce the violation. The CloseResource rule behaves correctly. I tested versions 6.29.0, 6.30.0, 6.31.0, with and without type resolution.
I can provoke a violation, if I set the rule property "closeAsDefaultTarget" to false - then "file.close()" is ignored and the rule rightly says, the file is not closed.

I've attached my sample project: [issue-2977-sample.zip](https://github.com/pmd/pmd/files/6050045/issue-2977-sample.zip) Could you please make it fail?

**Comment by djh82** [comment]
@adangel OK, I've tested with 6.32.0, and I still get the same issue.  I think this should fail:

```java
public class FalsePositive {
    private boolean isHundredBytes = false;

    public void bar() throws Exception {
        RandomAccessFile file = null;
        try {
            file = new RandomAccessFile("name", "r");
            this.isHundredBytes = file.length() == 100;
        } finally {
            file.close();
        }
    }
}
```

**Comment by adangel** [comment]
@djh82 Thanks for the updates sample, now we are getting somewhere: Updated test project: 
[issue-2977-sample.zip](https://github.com/pmd/pmd/files/6063798/issue-2977-sample.zip)

The reported violation:

```
[INFO] PMD Failure: AnotherFalsePositive:10 Rule:CloseResource Priority:3 'file' is reassigned, but the original instance is not closed.
```

It's present since 6.32.0, 6.31.0, 6.30.0, 6.29.0. It's not reproducible with 6.28.0.

The line, where we display the violation, is actually this one:
```
this.isHundredBytes = file.length() == 100;
```
which doesn't make much sense - neither file is declared here nor it is assigned here....
